lib/rucio/core/rule.py

```diff
@@ -2892,13 +2893,16 @@ def __create_rule_approval_email(rule, session=None):
         free_space_after = 'undefined'
 
         try:
-            for usage in get_rse_usage(rse=target_rse, session=session):
-                if usage['source'] == 'storage':
-                    free_space = sizefmt(usage['free'])
-                    if did['bytes'] is None:
-                        free_space_after = 'undefined'
-                    else:
-                        free_space_after = sizefmt(usage['free'] - did['bytes'])
+            replicas = rucio.core.replica.list_replicas(dids=[{'scope': rule.scope,
```

Comment:
This is really a quite heavy query, since you are reading/streaming all rows from the DB to the server, but you are actually only interested in an aggregate. For big datasets/containers, this is an issue.
What you can do is if `did['bytes'] is None`, query with `get_did(..., dynamic=True)`
https://github.com/rucio/rucio/blob/ba88eaf0f1de3c8cd6c44546568c6aafdfa784b8/lib/rucio/core/did.py#L1094

Reply:
What `get_did` does it also exposes the `bytes` value of the DID. If the value is `None` though, it would reply None unless you use the `dynamic=True` option, which calculates the bytes on demand, however in a much more efficient way than listing the replicas explicitly.

Reply:
I’m sorry, I might have not described the motivation enough. What is needed in this particular case is to calculate the portion of a DID that is already available on the destination RSE. I will elaborate in the issue.

Reply:
Ah, I see, should have read the code more carefully 😄 
I understand now, yes like this it doesn't work like that, but list_replicas is still not the right approach for this, the aggregation should be done on the database side. Listing 30k+ replicas will take way too long and slow a simple rule insertion (for approval) down a lot. So maybe a new method to do this would be appropriate.